static/app/components/timeline/index.tsx

```diff
@@ -0,0 +1,177 @@
+import {Tooltip} from 'sentry/components/tooltip';
+import {space} from 'sentry/styles/space';
+import {defined} from 'sentry/utils';
+import type {Color} from 'sentry/utils/theme';
+
+export interface ColorConfig {
+  primary: Color;
+  secondary: Color;
+}
+
+export interface ItemProps {
+  icon: React.ReactNode;
+  timeString: string;
+  title: React.ReactNode;
+  children?: React.ReactNode;
+  colorConfig?: ColorConfig;
+  isActive?: boolean;
+  onClick?: React.MouseEventHandler<HTMLDivElement>;
+  onMouseEnter?: React.MouseEventHandler<HTMLDivElement>;
+  onMouseLeave?: React.MouseEventHandler<HTMLDivElement>;
+  startTimeString?: string;
+}
+
+export function Item({
+  title,
+  children,
+  icon,
+  timeString,
+  startTimeString,
+  colorConfig = {primary: 'gray300', secondary: 'gray200'},
+  isActive = false,
+  onClick,
+  onMouseEnter,
+  onMouseLeave,
+}: ItemProps) {
+  const hasRelativeTime = defined(startTimeString);
+  const placeholderTime = useMemo(() => new Date().toTimeString(), []);
+  const {
+    displayTime,
+    date,
+    timeWithMilliseconds: preciseTime,
+  } = hasRelativeTime
+    ? getFormattedTimestamp(timeString, startTimeString, true)
+    : getFormattedTimestamp(timeString, placeholderTime);
+
+  return (
+    <Row
+      color={colorConfig.secondary}
+      hasLowerBorder={isActive ?? false}
+      onClick={onClick}
+      onMouseEnter={onMouseEnter}
+      onMouseLeave={onMouseLeave}
+    >
+      <IconWrapper colorConfig={colorConfig} hasIconBorder={isActive ?? false}>
+        {icon}
+      </IconWrapper>
+      <Title color={colorConfig.primary}>{title}</Title>
+      <Timestamp>
+        <Tooltip title={`${preciseTime} - ${date}`}>{displayTime}</Tooltip>
+      </Timestamp>
+      <Spacer hasLine={isActive ?? false} />
+      <Content>{children}</Content>
+    </Row>
+  );
+}
+
+interface GroupProps {
+  children: React.ReactNode;
+}
+
+export function Group({children}: GroupProps) {
+  return <GroupWrapper>{children}</GroupWrapper>;
+}
+
+const GroupWrapper = styled('div')`
+  position: relative;
+  /* vertical line connecting items */
+  &::before {
+    content: '';
+    position: absolute;
+    left: 10.5px;
+    width: 1px;
+    top: 0;
+    bottom: 0;
+    background: ${p => p.theme.border};
+  }
+`;
+
+const Row = styled('div')<{color: string; hasLowerBorder: boolean}>`
+  position: relative;
+  color: ${p => p.theme.subText};
+  display: grid;
+  align-items: center;
+  grid-template: auto auto / 22px 1fr auto;
+  grid-column-gap: ${space(1)};
+  border-bottom: 1px solid ${p => (p.hasLowerBorder ? p.theme[p.color] : 'transparent')};
+  margin: ${space(1)} 0;
+  &:first-child {
+    margin-top: 0;
+  }
+  &:last-child {
+    margin-bottom: 0;
+    background: ${p => p.theme.background};
+  }
+  &:last-child > :last-child {
+    margin-bottom: ${p => (p.hasLowerBorder ? space(1) : 0)};
+  }
+`;
+
+const IconWrapper = styled('div')<{colorConfig: ColorConfig; hasIconBorder: boolean}>`
```

Comment:
since these are rendered 100x in a list i'd pull the colors into inline `style={{}}`

Reply:
Inlined a bunch of the styles for these components, thanks 👍